src/aks-preview/azext_aks_preview/managed_cluster_decorator.py

```diff
@@ -4355,77 +4352,156 @@ def update_app_routing_profile(self, mc: ManagedCluster) -> ManagedCluster:
-                                    scope=dns_zone,
-                                ):
-                                    logger.warning(
-                                        'Could not create a role assignment for App Routing. '
-                                        'Are you an Owner on this subscription?')
-                        except Exception as ex:
-                            raise CLIError('Error in granting dns zone permisions to managed identity.\n') from ex
-            else:
-                raise CLIError('App Routing must be enabled to modify DNS zone resource IDs.\n')
+            self._update_dns_zone_resource_ids(mc, dns_zone_resource_ids)
 
         return mc
 
+    def _enable_keyvault_secret_provider_addon(self, mc: ManagedCluster) -> None:
+        """Helper function to enable keyvault secret provider addon for the ManagedCluster object.
+
+        :return: None
+        """
+        mc.addon_profiles = mc.addon_profiles or {}
+        if not mc.addon_profiles.get(CONST_AZURE_KEYVAULT_SECRETS_PROVIDER_ADDON_NAME):
+            mc.addon_profiles[
+                CONST_AZURE_KEYVAULT_SECRETS_PROVIDER_ADDON_NAME
+            ] = self.models.ManagedClusterAddonProfile(  # pylint: disable=no-member
+                enabled=True,
+                config={
+                    CONST_SECRET_ROTATION_ENABLED: "false",
+                    CONST_ROTATION_POLL_INTERVAL: "2m",
+                },
+            )
+        elif not mc.addon_profiles[CONST_AZURE_KEYVAULT_SECRETS_PROVIDER_ADDON_NAME].enabled:
+            mc.addon_profiles[CONST_AZURE_KEYVAULT_SECRETS_PROVIDER_ADDON_NAME].enabled = True
+            mc.addon_profiles[CONST_AZURE_KEYVAULT_SECRETS_PROVIDER_ADDON_NAME].config = {
+                CONST_SECRET_ROTATION_ENABLED: "false",
+                CONST_ROTATION_POLL_INTERVAL: "2m",
+            }
+
+    def _attach_keyvault_id(self, mc: ManagedCluster, keyvault_id) -> None:
+        """Helper function to attach keyvault id to app routing addon.
+
+        :return: None
+        """
+
+        from msrestazure.tools import parse_resource_id
+        from azure.cli.command_modules.keyvault.custom import set_policy
+        from azext_aks_preview._client_factory import get_keyvault_client
+
+        if not is_valid_resource_id(keyvault_id):
+            raise InvalidArgumentValueError("Please provide a valid keyvault ID")
+
+        self.cmd.command_kwargs['operation_group'] = 'vaults'
+        keyvault_params = parse_resource_id(keyvault_id)
+        keyvault_subscription = keyvault_params['subscription']
+        keyvault_name = keyvault_params['name']
+        keyvault_rg = keyvault_params['resource_group']
+        keyvault_client = get_keyvault_client(self.cmd.cli_ctx, subscription_id=keyvault_subscription)
+        keyvault = keyvault_client.get(resource_group_name=keyvault_rg, vault_name=keyvault_name)
+        managed_identity_object_id = mc.ingress_profile.web_app_routing.identity.object_id
```

Comment:
I think the purpose to put this role assignment after MC created is to obtain identity information. But it seems that this is no longer the case after the change. Can it still get the object_id here?

Reply:
Thanks for catching that. I've moved the code for role assignment to the `postprocessing_after_mc_created` method so that the `object_id` can be retrieved from the updated managed cluster.